lib/db/observed.go

```diff
@@ -0,0 +1,60 @@
+// Copyright (C) 2020 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package db
+
+import (
+	"time"
+
+	"github.com/syncthing/syncthing/lib/protocol"
+)
+
+func (db *Lowlevel) AddOrUpdatePendingDevice(device protocol.DeviceID, name, address string) {
+	//FIXME locking? m.mut.Lock()
+	//FIXME locking? defer m.mut.Unlock()
+
+	key := db.keyer.GeneratePendingDeviceKey(nil, device[:])
+	timestamp := time.Now().Round(time.Second)
+	od := ObservedDevice{
+		Time:    &timestamp,
+		Name:    name,
+		Address: address,
+	}
+	bs, err := od.Marshal()
+	if err == nil {
+		err = db.Put(key, bs)
+	}
+	if err != nil {
+		l.Warnf("Failed to store pending device entry: %v", err)
+	}
+}
+
+func (db *Lowlevel) PendingDevices() (map[protocol.DeviceID]ObservedDevice, error) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingDevice})
+	if err != nil {
+		return nil, err
+	}
+	res := make(map[protocol.DeviceID]ObservedDevice)
+	defer iter.Release()
+	for iter.Next() {
+		bs, err := db.Get(iter.Key())
+		if err != nil {
+			return nil, err
+		}
+		deviceID := db.keyer.DeviceFromPendingDeviceKey(iter.Key())
+		var od ObservedDevice
+		err = od.Unmarshal(bs)
+		if len(deviceID) != protocol.DeviceIDLength || err != nil {
```

Comment:
The id length check isn't necessary, it's always valid coming from the keyer (or we would have huge problems).

Reply:
That's not true for this specific function. It just returns anything after the KeyType prefix, which might have a wrong length if the DB is corrupted, which is handled by removing the entry.

Reply:
The other `DeviceFrom...` functions do the same thing. If we expect everything coming from the db to be corrupted, we'd need to add such tests everywhere - that's simply not practical.

Reply:
That's because the other functions validate by looking up a ShortID in the index.  Since we are handling yet unknown devices here, they cannot be in the `deviceIdx` yet.

If you'd like, I can do the check here.  I just copied the approach from `NameFromDeviceFileKey()` and others.

Reply:
I don't think that lookup is intended as a validation: For known devices we store an int, not byte-slice, thus need to look up the device.

In any case if we deem the check necessary, `DeviceIDFromBytes` should be changed to return an error instead of panicking then. However I still do not believe this is necessary.

> If you'd like, I can do the check here. I just copied the approach from `NameFromDeviceFileKey()` and others.

I don't understand what you want to move.

Reply:
My point was that the index lookup can already generate an error, which is passed up through the keyer functions.  The other "extracting" functions which just return an appropriate sub-slice cannot return an error, so I thought that validating key component sizes was not **desired** at this level.

To avoid a panic, I check for the length before *using* the extracted key component, but I agree that it's not a nice place to perform the check.  It's the right place to recover from it, though (dropping the unusable DB entry).

So we could centralize the length check  in `DeviceFromPendingDeviceKey()` and add an error return value.  That would introduce a new dependency from `lib/db/keyer.go` to the `lib/protocol` package though, which I tried to avoid.  And it would be much more than the other functions do.

> In any case if we deem the check necessary, `DeviceIDFromBytes` should be changed to return an error instead of panicking then. However I still do not believe this is necessary.

I agree on avoiding the panic, but there will be some calling places to adjust.  Whether the check is necessary at all, I would answer from my experience: Starting with this PR, I changed the stored ID format mid-way and had invalid old data in the DB which lead to a hard-to-diagnose panic that cost me some hours.  The check is not expensive and good practice for defensive programming IMO.

Reply:
> Whether the check is necessary at all, I would answer from my experience: Starting with this PR, I changed the stored ID format mid-way and had invalid old data in the DB which lead to a hard-to-diagnose panic that cost me some hours.

In my opinion this is exactly a case where a panic is appropriate: This is a condition that must never occur in production. If it does, we shouldn't just drop the entry from db and carry on. This will likely just induce a rats-tail of other inconsistencies in the database.

And currently any error in the db will result in a panic further up in the call tree, thus at a more generic and less clear location in the code (i.e. backtrace will be less informative).

Reply:
So I learned that in Go, errors should be checked and handled right away.  That vibes well with my own opinion that more than half the effort in programming is not about making it work, but about making it sensibly handle failures (and report for diagnosis).

The panic actually led to *no error message at all* in the logs, especially not a stack trace, but only an empty API response.  That's why I added the check and logging in `observed.go` as an improvement.  So that's the "reporting" part - please outline if you have other ideas.

For the "handling" part, all checks I added would lead to problems later on, e.g. when the API returns invalid values.  I think that's what you meant with "higher up the call tree".  Observed devices / folders is not very valuable information because it will be added again at the next occasion when the remote connects again (if that doesn't happen, syncing would never happen either).  So any cached entries that we cannot interpret *right now*, we will *never* be able to use / repair / whatever.  Those two arguments led me to the conclusion that dropping the entry would be wisest.

Reply:
I think we should focus on the handling part in context of the db, I full agree on your general points about failures.

TLDR:  
I like handling errors, I don't argue against that. What is still open for me is:
1. It should happen consistently. Which to me means returning an error from `DeviceIDFromBytes`. Otherwise I'd need to understand how this instance is different from all the other such that panicking is appropriate in all other cases but not here (and even then I'd argue we should return an error and then explicitly panic in the other cases).
2. Unless you have concrete failure modes in mind, the only one left I know of is hardware corruption. And I don't want to hide db corruption by dropping the one corrupt part detected. When detecting db corruption either we can guarantee everything else is intact or we should stop working with that db.

Longer:

Some background about my bias towards not handling failures by just dropping the entry:  
That stems from dealing with lots of database errors where the origin is not known. Thus my ideal scenario would be that the db is always either in a fully consistent, uncorrupted state, or entirely unusable (aka Syncthing refuses to run with it). Recent changes in that direction are to add migrations or periodic clean up jobs for known, historic failure modes (and admittedly also a few known existing failures, without knowing how the happened originally - that's why they exist periodically), while the failure modes themselves are fixed. 

Now the question is why would you get a slice with an incorrect length or why would unmarshaling fail? If you know why and can mitigate that, I am fine with that. However the way I see it now, we have no scenario except hardware corruption where that would occur. And if we detect a db corrupted by hardware, I don't want to hide that. In my preference we either do a full consistency check to make sure this is the only corruption and fix it, or we fail "hard". Currently in the context of the db, that means panic, though that's not ideal and worked on. However when we do not panic anymore, we still should stop Syncthing and signify to the user that there is db corruption - not carry on.

Reply:
> 1. It should happen consistently. Which to me means returning an error from `DeviceIDFromBytes`. Otherwise I'd need to understand how this instance is different from all the other such that panicking is appropriate in all other cases but not here (and even then I'd argue we should return an error and then explicitly panic in the other cases).

It's different in that I didn't write all the other instances ;-)  Looking through all uses of this function, there seem to be more places where a "corrupt" DB would lead to an immediate panic.  With my very limited knowledge of the codebase, I can't say whether all of these are handled sanely, or if there even is any better way to handle them.  Many are just converting `[]byte`s to a `DeviceID` for formatted output.  On the other hand, I assume there are many other cases where invalid data in the DB would NOT lead to a panic, but garbled output or worse.

Making this case a little bit special, as I explained, is that pending devices are merely persisted to redisplay the notification anytime after restarts.  Dropping them hurts nothing, but avoids logging any corrupt entries over and over.

> 2. Unless you have concrete failure modes in mind, the only one left I know of is hardware corruption. And I don't want to hide db corruption by dropping the one corrupt part detected. When detecting db corruption either we can guarantee everything else is intact or we should stop working with that db.

That's actually something I would expect the underlying database backend to detect and report.  Opening a corrupt database should lead to total failure at *that time* instead of returning a single corrupt data byte *when it's read*, making the application notice the corruption at a "random" point in time.  I don't know if LevelDB fulfils that expectation.

So in light of many functions possibly returning an "error" object, what to do then if anything goes awry?

* `DeviceIDFromBytes()` does not, but panics.  No way to even log an error message after that point, which is why I struggled to find the cause at all.  The HTTP REST response "thread" (or whatever invokes the handler) just dies **silently**.

* `Unmarshal()` returns lots of possible errors, but at least they can be accessed and logged.

* My own checks for key component lengths are for things that never happen as long as Syncthing does not put messed-up entries in the DB:  An empty folder ID, could be checked already in `FolderFromPendingFolderKey()` which then returns an error instead.  In fact if the key is too short overall, that function would currently panic as well, caused by an out of bounds slice index: `key[1:0]` for example.  Most other "simple" keyer functions cannot return an error though, so it would be slightly inconsistent there.

* For the device length check, moving it to the keyer creates a dependency to the `protocol` package to know what the correct length is.  I'm in no position to judge if proliferating inter-package dependencies are acceptable.  In my own software projects I strictly forbid cyclic dependencies for example, limiting imports (includes) to the absolute minimum.

> Some background about my bias towards not handling failures by just dropping the entry:
> [...]

Sounds perfectly reasonable.  So overall, your suggested course of action would be to (slightly exaggerated and please correct me if I misunderstood):

1. Gracefully detect errors, using error return values where needed instead of panicking.  Changing `DeviceIDFromBytes()` to not panic would be needed if we want to drop the extra precautious check here.
2. Log a warning / error message to inform (whom?) about the bug / corruption.
3. Keep the entry for later forensic analysis by a developer.  Note that this might lead to flooding the log as the previous steps are carried out anytime the entry is iterated over.
4. Check and repair the DB if possible to detect other corruptions.
5. Possibly end the program gracefully after encountering such an invalid entry, assuming the whole DB might be unusable.  Is that really necessary because of a potential future bug that wrote an invalid DB entry sometime, meant only as a cache?

> Now the question is why would you get a slice with an incorrect length or why would unmarshaling fail? If you know why and can mitigate that, I am fine with that. However the way I see it now, we have no scenario except hardware corruption where that would occur. [...]

Whole DB corruption is not what I was expecting.  More if for whatever reason a **future** code change messes up key or value generation as they are put into the DB, my code should fail gracefully, with a severity appropriate for the importance of the failed entry.  I cannot mitigate failure causes now which might arise from future changes, no matter how hard I think about all corner cases now.  Isn't that the heart of defensive programming techniques?

Sorry for the overly long post, hope it still kind of makes sense...

Reply:
Wow, thanks a lot for the comprehensive reasoning!

I'll just quote excerpts to limit the length, I doubt anyone else will read all of this, and if some one does, the full think is just above :)

> It's different in that I didn't write all the other instances ;-)

I get that of course. My opinion is that it doesn't really make sense to add a check in one place (here) and ignore it in the others. I.e. this argument is neither for or against the check, it's just that I'd either do it everywhere or nowhere.

>  I don't know if LevelDB fulfils that [detect db corruption on open] expectation.

It doesn't, at least not in all cases. Not sure whether that's part of the design, but practically from experience it's a clear no.

> * For the device length check, moving it to the keyer creates a dependency to the `protocol` package to know what the correct length is.

Keyer is not a package, it's part of db and that already depends on protocol (virtually everything does). However from my side there's no reason to move anything there, I'd either return an error from protocol, or keep the current special double checking or do nothing at all about length.



> Sounds perfectly reasonable. So overall, your suggested course of action would be to [points follow addressed below]

Generally this is probably way out of scope for this PR (as in a lot of work). I'll respond anyway but just to clarify: I do not expect you to act on it.

1. Yes
2 - 5. That would be part of a bigger refactor of the db - currently there's not good error handling in place. Effectively in my opinion we should mark the db as corrupted, stop syncthing and log an error about db corruption on a future start before exiting again.
5. Question in there together with the last part below

> [...] my code should fail gracefully, with a severity appropriate for the importance of the failed entry. [...]

Totally, emphasizing the "importance of the failed entry". That was my mistake in this exchange: Just because this is stored in the db doesn't mean it is as vital as the index info in the db.

So to summarize and hopefully go back to more productive exchanges (I did enjoy this exchange, but it was probably quite a lot longer than it's importance would require :) ):  
I do agree that dropping the entry in this case is appropriate. I would prefer for the length check to happen in `DeviceIDFromBytes` (i.e. return an error there instead of pancking), but I do not expect it in this PR and am ok with the duplicated check.

Reply:
Okay, so I misunderstood that you wanted to put the check in keyer.go.  Long story short, the current approach is good enough to not depend on a proper all-around-the-place fix. I'll add a FIXME though to indicate that check should be handled differently, long-term.